File: String/125. Valid Palindrome/125. Valid Palindrome.py
'''
Number: 125
Topic: Valid Palindrome
Date: 2021/10/4
Rate: Easy
https://leetcode.com/problems/valid-palindrome/

Given a string s, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Constraints:
* 1 <= s.length <= 2 * 105
* s consists only of printable ASCII characters.
'''

class Solution:
    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        i, j = 0, n-1
        while i < j:
            while i < j and not s[i].isalnum():
                i += 1
            while i < j and not s[j].isalnum():
                j -= 1
            if i < j and s[i].lower() != s[j].lower():
                return False
            i += 1; j -= 1
        return True

File: Archive/categorized-by-date/21-6-1/125. Valid Palindrome/125. Valid Palindrome.cpp
/*
Number: 125
Topic: Valid Palindrome 
Date: 2021/6/1
Rate: Easy
https://leetcode.com/problems/valid-palindrome/


Given a string s, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Constraints:
* 1 <= s.length <= 2 * 10^5
* s consists only of printable ASCII characters.
 
*/

class Solution {
public:
    bool isPalindrome(string s) {
        int i = 0; j = s.size()-1;
        while (i < j) {
            while (!isalnum(s[i]) && i < j)
                i++;
            while (!isalnum(s[j]) && i < j)
                j--;
            if (toupper(s[i]) != toupper(s[j]))
                return false;        
        }
        return true;
    }
};